contrib/endpoints/src/api_manager/service_control/aggregated.cc

```diff
@@ -341,13 +439,79 @@ Status Aggregated::GetStatistics(Statistics* esp_stat) const {
   return Status::OK;
 }
 
+template <class RequestType>
+const std::string& Aggregated::GetApiReqeustUrl() {
+  if (typeid(RequestType) == typeid(CheckRequest)) {
+    return url_.check_url();
+  } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
+    return url_.quota_url();
+  } else {
+    return url_.report_url();
+  }
+}
+
+template <class RequestType>
+int Aggregated::GetHttpRequestTimeout() {
+  int timeout_ms = 0;
+
+  // Set timeout on the request if it was so configured.
+  if (typeid(RequestType) == typeid(CheckRequest)) {
+    timeout_ms = kCheckDefaultTimeoutInMs;
+  } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
+    timeout_ms = kAllocateQuotaDefaultTimeoutInMs;
+  } else {
+    timeout_ms = kReportDefaultTimeoutInMs;
+  }
+
+  if (server_config_ != nullptr &&
+      server_config_->has_service_control_config()) {
+    const auto& config = server_config_->service_control_config();
+    if (typeid(RequestType) == typeid(CheckRequest)) {
+      if (config.check_timeout_ms() > 0) {
+        timeout_ms = config.check_timeout_ms();
+      }
+    } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
+      if (config.quota_timeout_ms() > 0) {
+        timeout_ms = config.quota_timeout_ms();
+      }
+    } else {
+      if (config.report_timeout_ms() > 0) {
+        timeout_ms = config.report_timeout_ms();
+      }
+    }
+  }
+
+  return timeout_ms;
+}
+
+template <class RequestType>
+const std::string& Aggregated::GetAuthToken() {
+  if (sa_token_) {
```

Comment:
not need to check sa_token_

return sa_token_->GetAuthToken( typeid(RequestType) == typeid(AllocateQuotaRequest)) ? x : y );